src/crypto/tls/delegated_credentials.go

```diff
@@ -0,0 +1,533 @@
+// plus ValidTime seconds. This function simply checks that the current time
+// ('now') is before the end of the validity interval.
+func (dc *DelegatedCredential) IsExpired(start, now time.Time) bool {
+	end := start.Add(dc.Cred.validTime)
+	return !now.Before(end)
+}
+
+// InvalidTTL returns true if the credential's validity period is longer than the
+// maximum permitted. This is defined by the certificate's notBefore field
+// ('start') plus the ValidTime, minus the current time ('now').
+func (dc *DelegatedCredential) InvalidTTL(start, now time.Time) bool {
+	return dc.Cred.validTime > (now.Sub(start) + dcMaxTTL).Round(time.Second)
+}
+
+// Credential stores the public components of a Delegated Credential.
+type Credential struct {
+	// The amount of time for which the credential is valid. Specifically, the
+	// the credential expires 'ValidTime' seconds after the 'notBefore' of the
+	// delegation certificate. The delegator shall not issue delegated
+	// credentials that are valid for more than 7 days from their current time.
+	//
+	// When this data structure is serialized, this value is converted to a
+	// uint32 representing the duration in seconds.
+	validTime time.Duration
+	// The signature scheme associated with the credential public key.
+	// This is expected to be the same as the CertificateVerify.algorithm
+	// sent by the server.
+	expCertVerfAlgo SignatureScheme
+	// The credential's public key.
+	PublicKey crypto.PublicKey
+}
+
+// DelegatedCredential stores the credential and its signature.
+type DelegatedCredential struct {
+	// The serialized form of the Credential
+	Raw []byte
+
+	// Credential stores the public components of a Delegated Credential
+	Cred *Credential
+
+	// The signature scheme used to sign the credential
+	Algorithm SignatureScheme
+
+	// The Credential's delegation: a signature that binds the credential to
+	// the end-entity certificate's public key
+	Signature []byte
+}
+
+// marshalPublicKeyInfo returns a DER encoded PublicKeyInfo
+// from a Delegated Credential (as defined in the X.509 standard).
+// The following key types are currently supported: *ecdsa.PublicKey
+// and ed25519.PublicKey. Unsupported key types result in an error.
+// rsa.PublicKey is not supported a defined by the draft.
+func (cred *Credential) marshalPublicKeyInfo() ([]byte, error) {
+	switch cred.expCertVerfAlgo {
+	case ECDSAWithP256AndSHA256,
+		ECDSAWithP384AndSHA384,
+		ECDSAWithP521AndSHA512,
+		Ed25519:
+		serPub, err := x509.MarshalPKIXPublicKey(cred.PublicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		return serPub, nil
+	default:
+		return nil, fmt.Errorf("tls: unsupported signature scheme: 0x%04x", cred.expCertVerfAlgo)
+	}
+}
+
+// unmarshalPublicKeyInfo parses a DER encoded PublicKeyInfo from a Delegated
+// Credential to a public key and its corresponding algorithm.
+func unmarshalPublicKeyInfo(serialized []byte) (crypto.PublicKey, SignatureScheme, error) {
```

Comment:
It feels like this function should be a more generic function in the tls module, but the other similar function `signatureSchemesForCertificate` is just a bit too different to be merged with this one.

Reply:
Yes. I agree. At some point, I was thinking on refactoring some of the tls functions but that will mean a bigger change for the standard lib rather than just DC.